modules/syslogformat/syslog-format.c

```diff
@@ -340,6 +363,16 @@ _syslog_format_parse_legacy_program_name(LogMessage *msg, const guchar **data, g
     {
       _skip_char(&src, &left);
     }
+
+  if ((flags & LP_CHECK_PROGRAM))
+    {
+      if (!_validate_program(prog_start, (src - prog_start)))
+        {
+          log_msg_set_tag_by_id(msg, LM_T_SYSLOG_RFC_3164_INVALID_PROGRAM);
+          return;
+        }
+    }
```

Comment:
This is one of the most performance-critical paths of AxoSyslog.
For this reason, can we reimplement `LP_CHECK_PROGRAM` similarly to how `LP_CHECK_HOSTNAME` is done?
(Avoiding the double iteration by moving the character validation inside the while loop above.)